File: backend/app/services/chunking.py
"""Page-aware recursive document chunking."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from langchain_text_splitters import RecursiveCharacterTextSplitter

from app.services.extraction import ExtractedSection


@dataclass(frozen=True, slots=True)
class TextChunk:
    """Chunk ready for relational and vector storage."""

    chunk_index: int
    page_number: int | None
    chunk_text: str
    token_count: int
    chunk_metadata: dict[str, Any]


class DocumentChunker:
    """Recursive splitter preserving page/heading provenance."""

    def __init__(self, chunk_size: int = 800, chunk_overlap: int = 150) -> None:
        self.splitter = RecursiveCharacterTextSplitter(
            chunk_size=chunk_size,
            chunk_overlap=chunk_overlap,
            length_function=len,
            separators=["\n\n", "\n", ". ", " ", ""],
        )

    def split(self, sections: list[ExtractedSection]) -> list[TextChunk]:
        """Split each source section independently so citations remain precise."""
        chunks: list[TextChunk] = []
        for section in sections:
            for text in self.splitter.split_text(section.text):
                clean_text = text.strip()
                if not clean_text:
                    continue
                chunks.append(
                    TextChunk(
                        chunk_index=len(chunks),
                        page_number=section.page_number,
                        chunk_text=clean_text,
                        token_count=max(1, len(clean_text) // 4),
                        chunk_metadata=section.metadata,
                    )
                )
        return chunks

File: backend/src/services/chunking.js
import { log, logProcessFinished, logProcessStarted } from "../logger.js";

function chooseBoundary(text, start, preferredEnd) {
  if (preferredEnd >= text.length) {
    return text.length;
  }
  const minimum = start + Math.floor((preferredEnd - start) * 0.55);
  for (const separator of ["\n\n", "\n", ". ", " "]) {
    const candidate = text.lastIndexOf(separator, preferredEnd);
    if (candidate >= minimum) {
      return candidate + separator.length;
    }
  }
  return preferredEnd;
}

export class DocumentChunker {
  constructor(chunkSize = 800, chunkOverlap = 150) {
    if (chunkOverlap >= chunkSize) {
      throw new Error("chunkOverlap must be smaller than chunkSize");
    }
    this.chunkSize = chunkSize;
    this.chunkOverlap = chunkOverlap;
  }

  split(sections) {
    logProcessStarted("Split document into chunks", {
      sections: sections.length,
      chunk_size: this.chunkSize,
      chunk_overlap: this.chunkOverlap,
    });
    const chunks = [];
    for (const section of sections) {
      const text = section.text.trim();
      let start = 0;
      while (start < text.length) {
        const end = chooseBoundary(text, start, Math.min(text.length, start + this.chunkSize));
        const chunkText = text.slice(start, end).trim();
        if (chunkText) {
          const chunk = {
            chunkIndex: chunks.length,
            pageNumber: section.pageNumber ?? null,
            chunkText,
            tokenCount: Math.max(1, Math.floor(chunkText.length / 4)),
            chunkMetadata: section.metadata ?? {},
          };
          chunks.push(chunk);
          log("info", "document_chunk_created", {
            message: "Document chunk created.",
            chunk_index: chunk.chunkIndex,
            page_number: chunk.pageNumber,
            chars: chunk.chunkText.length,
            estimated_tokens: chunk.tokenCount,
            metadata: chunk.chunkMetadata,
            preview: chunk.chunkText,
          });
        }
        if (end >= text.length) {
          break;
        }
        const nextStart = Math.max(start + 1, end - this.chunkOverlap);
        start = nextStart;
      }
    }
    logProcessFinished("Split document into chunks", { chunks: chunks.length });
    return chunks;
  }
}
